sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileClient.java

```diff
@@ -2059,8 +2064,41 @@ public ShareFileUploadInfo upload(InputStream data, long length, ParallelTransfe
      */
     public Response<ShareFileUploadInfo> uploadWithResponse(ShareFileUploadOptions options,
         Duration timeout, Context context) {
-        return StorageImplUtils.blockWithOptionalTimeout(
-            shareFileAsyncClient.uploadWithResponse(options, context), timeout);
+        StorageImplUtils.assertNotNull("options", options);
+        ShareRequestConditions validatedRequestConditions = options.getRequestConditions() == null
+            ? new ShareRequestConditions()
+            : options.getRequestConditions();
+        final ParallelTransferOptions validatedParallelTransferOptions =
+            ModelHelper.populateAndApplyDefaults(options.getParallelTransferOptions());
+        long validatedOffset = options.getOffset() == null ? 0 : options.getOffset();
+        int chunkSize = (int) Math.min(validatedParallelTransferOptions.getBlockSizeLong(), Integer.MAX_VALUE);
+
+        Callable<Response<ShareFileUploadInfo>> operation = () -> {
```

Comment:
Few changes made here don't align with the previous design:

- Old design attempted to upload in a single upload if the max put size limit wasn't crossed (4MB). The test failures are due to this change as this is always uploading in chunks, ignoring an incorrect size mismatch between what the stream contains and the stated length.
- This doesn't upload in parallel if there is more than 4MB of data contained in the stream.